indexer/pkg/indexer/indexer.go

```diff
@@ -0,0 +1,578 @@
+	if err != nil {
+		return fmt.Errorf("failed to get last backward indexed block: %w", err)
+	}
+	bi.logger.Info("last indexed block", "blockNumber", lastBackwardIndexedBlock, "direction", "backward")
+
+	if lastBackwardIndexedBlock == nil || lastBackwardIndexedBlock.Cmp(big.NewInt(0)) == 0 {
+		bi.lastBackwardIndexedBlock = new(big.Int).SetUint64(latestBlockNumber - 1)
+	} else {
+		bi.lastBackwardIndexedBlock = new(big.Int).Sub(lastBackwardIndexedBlock, big.NewInt(1))
+	}
+
+	return nil
+}
+
+func (bi *BlockchainIndexer) fetchForwardBlocks(ctx context.Context) {
+	ticker := time.NewTicker(bi.indexInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			latestBlockNumber, err := bi.ethClient.BlockNumber(ctx)
+			if err != nil {
+				bi.logger.Error("Failed to get latest block number", "error", err)
+				continue
+			}
+
+			for blockNum := new(big.Int).Add(bi.lastForwardIndexedBlock, big.NewInt(1)); blockNum.Cmp(latestBlockNumber) <= 0; blockNum.Add(blockNum, big.NewInt(5)) {
+				endBlockNum := new(big.Int).Add(blockNum, big.NewInt(4))
+				if endBlockNum.Cmp(latestBlockNumber) > 0 {
+					endBlockNum.Set(latestBlockNumber)
+				}
+
+				blockNums := []*big.Int{}
+				for bn := new(big.Int).Set(blockNum); bn.Cmp(endBlockNum) <= 0; bn.Add(bn, big.NewInt(1)) {
+					blockNums = append(blockNums, new(big.Int).Set(bn))
+				}
+
+				blocks, err := bi.fetchBlocks(ctx, blockNums)
+				if err != nil {
+					bi.logger.Error("Failed to fetch blocks", "start", blockNum, "end", endBlockNum, "error", err)
+					continue
+				}
+
+				for _, block := range blocks {
+					bi.blockChan <- block
+					bi.lastForwardIndexedBlock.Set(block.Number())
+				}
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) fetchBackwardBlocks(ctx context.Context) {
+	ticker := time.NewTicker(bi.indexInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if bi.lastBackwardIndexedBlock.Sign() < 0 {
+				return
+			}
+			block, err := bi.ethClient.BlockByNumber(ctx, bi.lastBackwardIndexedBlock)
+			if err != nil {
+				bi.logger.Error("Failed to get block", "number", bi.lastBackwardIndexedBlock, "error", err)
+				continue
+			}
+			bi.blockChan <- block
+			bi.lastBackwardIndexedBlock.Set(new(big.Int).Sub(bi.lastBackwardIndexedBlock, big.NewInt(1)))
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) processBlocks(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case block := <-bi.blockChan:
+			if err := bi.indexBlock(ctx, block); err != nil {
+				bi.logger.Error("Failed to index block", "error", err)
+			}
+			if err := bi.indexTransactions(ctx, block); err != nil {
+				bi.logger.Error("Failed to index transactions", "error", err)
+			}
+		}
+	}
+}
+
+func (bi *BlockchainIndexer) indexBlock(ctx context.Context, block *types.Block) error {
+	timestamp := time.UnixMilli(int64(block.Time())).UTC().Format("2006-01-02T15:04:05.000Z")
```

Comment:
This time layout `2006-01-02T15:04:05.000Z` is used multiple times, extract it to a constant.